src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +257,300 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+	}
+
+	fclose(txnBufferForRead);
+
+	return true;
+}
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file when we encounter
+ * a COMMIT without a valid txnCommitLSN.
+ *
+ * This ensures that the apply process doesn't block the generation of the
+ * transaction metadata file created by the transform process on arrival of
+ * a COMMIT message.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	LogicalMessageMetadata metadata = { 0 };
+
+	if (!parseSQLAction((char *) line, &metadata))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* We don't use this metadata in this function */
+	free(metadata.jsonBuffer);
+
+
+	if (metadata.action == STREAM_ACTION_BEGIN)
+	{
+		/*
+		 * Enable transaction buffering only when the BEGIN doesn't have
+		 * valid commit LSN, otherwise we can skip buffering and stream
+		 * the transaction directly.
+		 */
+		replayCtx->skipTxnBuffering = metadata.txnCommitLSN != InvalidXLogRecPtr;
+		if (replayCtx->skipTxnBuffering)
+		{
+			return stream_replay_line(ctx, line, stop);
+		}
+
+		if (replayCtx->isTxnBuffering)
+		{
+			/*
+			 * When the follow switches from prefetch to replay mode, we
+			 * call stream_transform_stream which might stream the partially
+			 * written transaction created during the prefetch mode.
+			 *
+			 * For example, lets consider the partially written C.sql file,
+			 *
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * KEEPALIVE;
+			 *
+			 * After switching to the replay mode, logical decoding will resume
+			 * from consistent point which again starts with a valid transcation
+			 * block.
+			 *
+			 * Lets assume the following contents streamed from the transform
+			 * to catchup process in UNIX PIPE after switching to replay mode,
+			 *
+			 * BEGIN -- {"xid": 999, "commitLSN": "0/1230"};
+			 * INSERT INTO A VALUES (1);
+			 * INSERT INTO A VALUES (2);
+			 * COMMIT -- {"xid": 999, "lsn": "0/1230"};
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * COMMIT -- {"xid": 1000, "lsn": "0/1234"};
+			 *
+			 * Contents of C.sql will be streamed by stream_transform_stream
+			 * and followed by the contents from UNIX PIPE. Since both of the
+			 * contents are streamed one after the other, the second block
+			 * contains the full content of previously written
+			 * transaction. So, we can skip previous transaction buffer and
+			 * start buffering from the current transaction.
+			 */
+			log_debug("Received %s when transaction is already "
+					  "in buffering mode, ignoring.", line);
```

Comment:
Please rewrite the message so that it ends with `": %s", line`. This makes it easier to process the logs later.